### ISSUE TITLE
`401 Bad credentials []` when installing plugin from GitHub (unless sudo is used)

### ISSUE THREAD

**Issue by eherot** [terraform-linters/tflint#1268]
Getting an inexplicable `401 Bad credentials []` error during init when installing a plugin which goes away if I use sudo:

Without sudo:
```console
❯ rm -rf ~/.tflint.d && tflint --init --loglevel trace -c ../.tflint.hcl
10:40:53 config.go:107: [INFO] Load config: ../.tflint.hcl
10:40:53 config.go:330: [DEBUG] Config loaded
10:40:53 config.go:331: [DEBUG]   Module: false
10:40:53 config.go:332: [DEBUG]   Force: false
10:40:53 config.go:333: [DEBUG]   IgnoreModules: map[string]bool{}
10:40:53 config.go:334: [DEBUG]   Varfiles: []string{}
10:40:53 config.go:335: [DEBUG]   Variables: []string{}
10:40:53 config.go:336: [DEBUG]   DisabledByDefault: false
10:40:53 config.go:337: [DEBUG]   PluginDir:
10:40:53 config.go:338: [DEBUG]   Rules: map[string]*tflint.RuleConfig{}
10:40:53 config.go:339: [DEBUG]   Plugins: map[string]*tflint.PluginConfig{"aws":(*tflint.PluginConfig)(0xc0003b9680)}
Installing `aws` plugin...
10:40:53 install.go:83: [DEBUG] Mkdir plugin dir: /Users/eherot/.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.9.0
10:40:53 install.go:156: [DEBUG] Request to https://api.github.com/repos/terraform-linters/tflint-ruleset-aws/releases/tags/v0.9.0
Failed to install a plugin. An error occurred:

Error: Failed to fetch GitHub releases: GET https://api.github.com/repos/terraform-linters/tflint-ruleset-aws/releases/tags/v0.9.0: 401 Bad credentials []
```
With sudo:
```console
❯ rm -rf ~/.tflint.d && sudo tflint --init --loglevel trace -c ../.tflint.hcl
Password:
10:41:30 config.go:107: [INFO] Load config: ../.tflint.hcl
10:41:30 config.go:330: [DEBUG] Config loaded
10:41:30 config.go:331: [DEBUG]   Module: false
10:41:30 config.go:332: [DEBUG]   Force: false
10:41:30 config.go:333: [DEBUG]   IgnoreModules: map[string]bool{}
10:41:30 config.go:334: [DEBUG]   Varfiles: []string{}
10:41:30 config.go:335: [DEBUG]   Variables: []string{}
10:41:30 config.go:336: [DEBUG]   DisabledByDefault: false
10:41:30 config.go:337: [DEBUG]   PluginDir:
10:41:30 config.go:338: [DEBUG]   Rules: map[string]*tflint.RuleConfig{}
10:41:30 config.go:339: [DEBUG]   Plugins: map[string]*tflint.PluginConfig{"aws":(*tflint.PluginConfig)(0xc000746300)}
Installing `aws` plugin...
10:41:30 install.go:83: [DEBUG] Mkdir plugin dir: /Users/eherot/.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.9.0
10:41:30 install.go:156: [DEBUG] Request to https://api.github.com/repos/terraform-linters/tflint-ruleset-aws/releases/tags/v0.9.0
10:41:31 install.go:163: [DEBUG] asset found: checksums.txt
10:41:31 install.go:163: [DEBUG] asset found: checksums.txt.sig
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_darwin_amd64.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_darwin_arm64.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_linux_386.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_linux_amd64.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_linux_arm.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_linux_arm64.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_windows_386.zip
10:41:31 install.go:163: [DEBUG] asset found: tflint-ruleset-aws_windows_amd64.zip
10:41:31 install.go:93: [DEBUG] Download checksums.txt
10:41:31 install.go:179: [DEBUG] Request to https://api.github.com/repos/terraform-linters/tflint-ruleset-aws/releases/assets/48776727
10:41:31 install.go:197: [DEBUG] Downloaded to /tmp/tflint-download-temp-file-2453786876
10:41:31 install.go:104: [DEBUG] Download checksums.txt.sig
10:41:31 install.go:179: [DEBUG] Request to https://api.github.com/repos/terraform-linters/tflint-ruleset-aws/releases/assets/48776937
10:41:31 install.go:197: [DEBUG] Downloaded to /tmp/tflint-download-temp-file-659414840
10:41:31 install.go:119: [DEBUG] Verified signature successfully
10:41:31 install.go:122: [DEBUG] Download tflint-ruleset-aws_darwin_amd64.zip
10:41:31 install.go:179: [DEBUG] Request to https://api.github.com/repos/terraform-linters/tflint-ruleset-aws/releases/assets/48776731
10:41:32 install.go:197: [DEBUG] Downloaded to /tmp/tflint-download-temp-file-4044717957
10:41:32 install.go:138: [DEBUG] Matched checksum successfully
10:41:32 install.go:213: [DEBUG] file found in zip: tflint-ruleset-aws
10:41:33 install.go:144: [DEBUG] Installed /Users/eherot/.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.9.0/tflint-ruleset-aws successfully
Installed `aws` (source: github.com/terraform-linters/tflint-ruleset-aws, version: 0.9.0)
```
My config:
```hcl
plugin "aws" {
  enabled = true
  version = "0.9.0"
  source  = "github.com/terraform-linters/tflint-ruleset-aws"
}
```

## Version
Without sudo:
```console
❯ tflint -v --enable-plugin=aws
TFLint version 0.33.1
+ ruleset.aws (0.8.0-bundled)
```
With sudo (just to show they are the same):
```console
❯ sudo tflint -v --enable-plugin=aws
TFLint version 0.33.1
+ ruleset.aws (0.8.0-bundled)
```

**Comment by bendrucker** [comment]
Are you supplying credentials for GitHub or making these requests anonymously? Can't see how root perms would have any impact here.

**Comment by eherot** [comment]
I'm not trying to, but I suppose it's possible that there may be something in my environment that's tripping it up. Not quite sure what to look for though. It would be helpful if the trace output showed the actual request...

**Comment by bendrucker** [comment]
`git clone` is doing different things than GitHub REST API calls. TFLint isn't cloning anything, it's fetching release artifacts from the API via REST.

`env | grep GITHUB_TOKEN` will tell you:

https://github.com/terraform-linters/tflint/blob/47ee5a45db833edac874d38020864568dc446317/docs/user-guide/plugins.md#avoiding-rate-limiting

**Comment by eherot** [comment]
Yep, that was it. I figured it might be something like that but it would be nice if there were enough output to see that that was the issue...

**Comment by bendrucker** [comment]
What config do you have exactly? And how is `sudo` affecting it? 

> it would be nice if there were enough output to see that that was the issue...

It would be helpful to have the logs indicate when a token has been supplied.

The APIs being accessed here allow anonymous access, since it's a public repo. The only way you're going to get a 401 is to supply invalid credentials. Supplying no credentials or valid credentials should work. You will still need to infer why/how your environment is supplying those invalid inputs.

**Comment by bendrucker** [comment]
Logs added. Forgot that `sudo` doesn't pass through environment variables by default.

**Comment by eherot** [comment]
Specifically I had a `GITHUB_TOKEN` set that (apparently) was not working to access GitHub. Doing `unset GITHUB_TOKEN` fixed the issue.
